storage/shard.go

```diff
@@ -532,96 +573,107 @@ func (s *dbShard) tickAndExpire(
 	tickSleepBatch := s.currRuntimeOptions.tickSleepSeriesBatchSize
 	tickSleepPerSeries := s.currRuntimeOptions.tickSleepPerSeries
 	s.RUnlock()
-	s.forEachShardEntry(func(entry *dbShardEntry) bool {
-		if i > 0 && i%tickSleepBatch == 0 {
-			// NB(xichen): if the tick is cancelled, we bail out immediately.
-			// The cancellation check is performed on every batch of entries
-			// instead of every entry to reduce load.
-			if c.IsCancelled() {
-				return false
+	s.forEachShardEntryBatch(func(currEntries []*dbShardEntry) bool {
+		var expired []*dbShardEntry
+		for _, entry := range currEntries {
+			if i > 0 && i%tickSleepBatch == 0 {
+				// NB(xichen): if the tick is cancelled, we bail out immediately.
+				// The cancellation check is performed on every batch of entries
+				// instead of every entry to reduce load.
+				if c.IsCancelled() {
+					return false
+				}
+				// NB(prateek): Also bail out early if the shard is closing,
+				// unless it's the final tick issued during the Close(). This
+				// final tick is required to release resources back to our pools.
+				if policy != tickPolicyCloseShard && s.isClosing() {
+					terminatedTickingDueToClosing = true
+					return false
+				}
+				// Throttle the tick
+				sleepFor := time.Duration(tickSleepBatch) * tickSleepPerSeries
+				s.sleepFn(sleepFor)
+				slept += sleepFor
 			}
-			// NB(prateek): Also bail out early if the shard is closing,
-			// unless it's the final tick issued during the Close(). This
-			// final tick is required to release resources back to our pools.
-			if policy != tickPolicyCloseShard && s.isClosing() {
-				terminatedTickingDueToClosing = true
-				return false
+
+			var (
+				result series.TickResult
+				err    error
+			)
+			switch policy {
+			case tickPolicyRegular:
+				result, err = entry.series.Tick()
+			case tickPolicyCloseShard:
+				err = series.ErrSeriesAllDatapointsExpired
 			}
-			// Throttle the tick
-			sleepFor := time.Duration(tickSleepBatch) * tickSleepPerSeries
-			s.sleepFn(sleepFor)
-			slept += sleepFor
-		}
-		var (
-			result series.TickResult
-			err    error
-		)
-		switch policy {
-		case tickPolicyRegular:
-			result, err = entry.series.Tick()
-		case tickPolicyCloseShard:
-			err = series.ErrSeriesAllDatapointsExpired
-		}
-		if err == series.ErrSeriesAllDatapointsExpired {
-			expired = append(expired, entry.series)
-			r.expiredSeries++
-			if len(expired) >= expireBatchLength {
-				// Purge when reaching max batch size to avoid large array growth
-				// and ensure smooth rate of elements being returned to pools.
-				// This method does not run using a lock so this is safe to
-				// perform inline.
-				s.purgeExpiredSeries(expired)
-				for i := range expired {
-					expired[i] = nil
+			if err == series.ErrSeriesAllDatapointsExpired {
+				expired = append(expired, entry)
+				r.expiredSeries++
+				if len(expired) >= expireBatchLength {
```

Comment:
As discussed, let's perhaps remove the limit of expiring in smaller batches, if we interrupt less that's probably better anyhow.

Reply:
sure thing

Reply:
done